src/Workspaces/Core/Portable/Shared/Utilities/ProducerConsumer.cs

```diff
@@ -102,18 +104,35 @@ internal static class ProducerConsumer<TItem>
         TArgs args,
         CancellationToken cancellationToken)
     {
-        return RunAsync(
-            // We're running in parallel, so we def have multiple writers
-            ProducerConsumerOptions.SingleReaderOptions,
-            produceItems: static (callback, args, cancellationToken) =>
-                RoslynParallel.ForEachAsync(
-                    args.source,
-                    cancellationToken,
-                    async (source, cancellationToken) =>
-                        await args.produceItems(source, callback, args.args, cancellationToken).ConfigureAwait(false)),
-            consumeItems: static (enumerable, args, cancellationToken) => args.consumeItems(enumerable, args.args, cancellationToken),
-            args: (source, produceItems, consumeItems, args),
-            cancellationToken);
+        // Bridge to sibling helper that operates on an IAsyncEnumerable.
+        return RunParallelAsync(source.AsAsyncEnumerable(), produceItems, consumeItems, args, cancellationToken);
```

Comment:
was able to get several overloads to be much simpler by having the iAsyncEnumerable forms be the primary ones, and the IEnumerable forms just trivially defer to that.